cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandler.java

```diff
@@ -74,6 +85,150 @@
     this.nsStore = nsStore;
   }
 
+  /**
+   * Returns user repository information
+   */
+  @Path("git/repo/info")
+  @GET
+  public void getRepoInfo(HttpRequest request, HttpResponder responder) throws Exception {
+    //check if repo info exists
+    if (userRepo.toString().equals("{}")) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, GSON.toJson("Please add repo information first."));
+    } else {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(userRepo));
+    }
+  }
+
+  @Path("git/repo/add")
+  @PUT
+  public void addRepoInfo(FullHttpRequest request, HttpResponder responder) throws Exception {
+    //first check if repository info has already been saved
+    if (!userRepo.toString().equals("{}")) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, GSON.toJson("Repo Information is already saved"));
+    } else {
+      JsonObject jObj = DECODE_GSON.fromJson(request.content().toString(StandardCharsets.UTF_8), JsonObject.class);
+      //check that the Json has the necessary fields
+      if (jObj.has("nickname") && jObj.has("url") && jObj.has("default_branch") && jObj.has("path")) {
+        String nickname = jObj.getAsJsonPrimitive("nickname").toString().replace("\"", "");
+        String defaultBranch = jObj.getAsJsonPrimitive("default_branch").toString().replace("\"", "");
+        String path = jObj.getAsJsonPrimitive("path").toString().replace("\"", "");
+
+        String giturl = jObj.getAsJsonPrimitive("url").toString().replace("\"", "");
```

Comment:
extract your url parsing logic to a small method or class and add enough comments to make it more readable.